[Pang Yuan Ker] iP

pyk595 opened the pull request:
# LarryBot
> "What Prof said about the mind creating ideas instead of storing them" ([Source](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html))

LarryBot allows you to ease some of the loneliness of being a CS student taking an online semester:upside_down_face:by allowing you to simulate some semblance of human interaction while you write up a task schedule. Some features include:
- Writing tasks
- Marking tasks as **done**/undone
- Deleting ~tasks~
- Finding tasks with a keyword
- Saving tasks to a text file

And it's **FREE**!

All you need to do is to:
1. Go download the [alpha](https://github.com/pyk595/ip/releases/download/A-Jar/Duke.jar)
2. Double click it
3. Add your tasks in CLI

The available commands are:
- [X] Creating a task item 
- [X] Marking a task item
- [X] Deleting a task item
- [x] Finding a task item
- [ ] Autosave (Wait for the updated version to be uploaded for optimal experience)

Use the command `list` to view your list in the app.

Other CLI inputs are:
```
todo <TaskName>
event <TaskName> /at <YYYY-MM-DD>
deadline <TaskName> /by <YYYY-MM-DD>
find <KeyWord>
delete <TaskNumber>
mark <TaskNumber>
unmark <TaskNumber>
```

The best part of this program is that you only have 2 lines of code for the launcher:
```Java
public static void main(String[] args) {
        CommandManager commander = new CommandManager();
        commander.run();
    }
```

Bye!

Commit:
Add Gradle support

pyk595 committed:
Level 1. Greet, Echo Exit

pyk595 committed:
Level 2. Add, List: added storing and display functionality, edited formatting for prints for aesthetic purposes

pyk595 committed:
Level 3. Mark as Done: Added marking and unmarking function for tasks

pyk595 committed:
Level 4. ToDos, Events, Deadlines: Refactored code for Task to apply polymorphism and added TaskManager class for more encapsulation

pyk595 committed:
Handle Errors: Created 3 new exception classes, handled 5 different error types

pyk595 committed:
Level 6. Delete: added functionality for deleting tasks from list. Handled potential errors arising from deletion

pyk595 committed:
Automated Text UI Testing: Removed unexpected blank spaces from prints, wrote test cases in input.txt and output in EXPECTED.TXT

pyk595 committed:
Added more test cases to check for invalid input

pyk595 committed:
Minor patch to delete function, added remaining tasks to prints

pyk595 committed:
Organize into Packages

Current classes are not in packages.
Adding classes into packages facilitate further developments in an OOP
manner, while ensuring that we improve encapsulation and abstraction.

pyk595 committed:
Refactor code into OOP style

Previously, the TaskManager class was used to parse and process commands.
As this does not optimally use the OOP paradigm, much of the code was
refactored to allow for the use of polymorphism and encapsulation.

This would allow the project to be more scalable for future increments.

pyk595 committed:
Add Enum for Type

Previously, a direct switch was used for tasks.
For a better typed set up, enums are used for the commands instead.

pyk595 committed:
Add save function

Previously, the project was only able to write and modify tasks in
the existing working list. To allow for users to save the task list,
the save function was introduced.

pyk595 committed:
Add LocalDate formatting

Previously, the event and deadline tasks represented dates with Strings.
To allow for more flexible use of the dates, dates were changed to
LocalDate objects instead

pyk595 committed:
Merge branch 'branch-Level-7'

Add save feature into master branch

pyk595 committed:
Merge branch 'branch-Level-8'

Add LocalDate formatting to master branch

pyk595 committed:
Patch on level 7

Previously, save command could only save for viewing in text file.
To allow for automatic saving and loading, text storage and text
loader was created.

Added a new InvalidOperationException for when program tries to
run mark on an existing marked task, and unmark on an unmarked
task.

Refactored exceptions such that most exceptions caught in execute
commands are caught in Command Manager instead.

pyk595 committed:
Refactor packages into duke

Previously, other packages were on the same level as duke.
Currently, other packages are placed in duke as well.

pyk595 committed:
Add JUnit tests

Previously, testing was only done using input and output comparison
via text-ui-test and manual regression test.

To test components such as Task objects and the UiManager object,
component tests were written for methods in these classes. Impure
methods are not tested by the component tests siince they are
tested by text-ui-test.

pyk595 committed:
Stash META-INF in gitignore

pyk595 committed:
Add find functionality using keywords

Previously, there was no find function to search for tasks using
keywords. This patch will add the find function and print a
formatted list containing all tasks with the keyword. Text-ui-test
is also updated to test the functionality of the feature.

pyk595 committed:
Add JavaDoc comments

Previously, no JavaDoc comments were used. To facilitate better
understanding of each method and class for documentation purposes,
JavaDoc comments are added to document all public methods and
classes.

pyk595 committed:
Style code according to Coding Standard

Previously, code was not styled fully to Coding Standard.
Edits are made such that code adheres to Coding Standard.

pyk595 committed:
Merge branch 'branch-A-CodingStandard'

Update master to adhere to coding standard.

pyk595 committed:
Merge branch 'branch-A-JavaDoc'

Add JavaDocs from parallel branch.
Handled conflicts during merge.

pyk595 committed:
Merge branch 'branch-Level-9'

Add find functionality from parallel branch.
Handled conflicts during merge.

pyk595 committed:
refactor: add storage folder into gitignore

pyk595 committed:
Refactor: Update save feature to use ser files

Update automatic save feature to work with ser files instead of
text files.

Add JavaDoc comments for ListLoader, ListStorage, SaveCommand and
FindCommand.

pyk595 committed:
fix: Add autosave upon modification commands

pyk595 committed:
Merge branch 'add-gradle-support'

Add gradle support in master branch

pyk595 committed:
Use Gradle

Automate project builds using Gradle.
Set up gradle to run Duke.
Set up gradle to run unit tests.

pyk595 committed:
Apply Checkstyle

Refactor: Format code to adhere to checkstyle guidelines

Tiffanylin21 commented on src/main/java/duke/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,88 @@
+            this.description = task;
+            break;
+        case DEADLINE:
+            String[] dInput = task.split("/by", 2);
```
```
Perhaps using a plural noun could be used for the variable name here?
```

Tiffanylin21 commented on src/main/java/duke/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,88 @@
+            this.date = dInput[1];
+            break;
+        case EVENT:
+            String[] eInput = task.split("/at", 2);
```
```
Perhaps consider using a plural noun here as well?
```

pyk595 replied:
Okay!

Tiffanylin21 commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * Constructs the DukeException Object
+     */
+    public DukeException() {
+    }
```
```
Perhaps consider collapsing these into 1 line?
```

pyk595 replied:
Okay!

Tiffanylin21 commented on src/main/java/duke/ui/ListLoader.java:
```diff
@@ -0,0 +1,40 @@
+    public void loadList() throws IOException, ClassNotFoundException {
+        FileInputStream readData = new FileInputStream("storage/list.ser");
+        ObjectInputStream readStream = new ObjectInputStream(readData);
+        ArrayList<Task> taskList = (ArrayList<Task>) readStream.readObject();
```
Perhaps consider using a plural noun as the name of the variable?

pyk595 replied:
Hmm I think I'll stick to tasklist since it keeps the purpose very clear on what it's supposed to hold, but thanks!

Tiffanylin21 commented on src/test/java/duke/tasks/DeadlineTest.java:
```diff
@@ -0,0 +1,65 @@
+
+import java.time.format.DateTimeParseException;
+
+import org.junit.jupiter.api.Test;
```
Perhaps consider grouping line 3 and 7 together as a block?

pyk595 replied:
I can't do that because of the static declaration. Doesn't pass checkstyle :C

Tiffanylin21 reviewed (commented):
Overall, I found your code clean and easy to read. I mostly just noted some minor variable-naming issues.

laughingkid-sg commented on src/main/java/duke/commands/CommandManager.java:
```diff
@@ -0,0 +1,135 @@
+                    break;
+                case "todo":
+                    AddTaskCommand todo = new AddTaskCommand(this.uiManager, this.taskManager, command[1], Type.TODO);
+                    todo.execute();
```
The execution command and save to list command perhaps could be abstracted to keep the switch block shorter. It seems that this method goes beyond the recommended 30 lines.

pyk595 replied:
Yeah it's a little lengthy, I'll take a look in a bit and try to refactor

laughingkid-sg commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    public String toString() {
+        if (isDone) {
+            return "[D]" + "[X] " + description + "(by: "
```
Could this part be improved? Maybe using String Format might better another reasonable option.

pyk595 replied:
Oh yeah that's true, thanks a bunch for the tip!

laughingkid-sg reviewed (commented):
LGTM. Just some nits.

pyk595 committed:
Create GUI for Duke

Previously the project used a CLI style print.
To make it more user friendly, a GUI was implemented using JavaFX.

pyk595 committed:
Merge branch 'branch-Level-10'

Add GUI commit to master branch

pyk595 committed:
Fix code based on PR comments

Task classes have a hard coded toString method.

The hard coding are error prone when expanding in future.

Changing to the String::format method allows us to use variable
input when dealing with task toString methods.

Let's use String::format for the toString method in task classes.

Using String::format is preferrable because future expansions may
require more variables and may potentially cause more errors if
hard coded.

pyk595 committed:
Use Assertions

Some assumptions were made in the code. By adding assertions, we
provide an additional layer of check at runtime to handle potential
errors.

Let's add assertions for invalid indices for numerical commands.

These assertions will ensure that indices less than 0 are handled.

pyk595 committed:
Merge pull request #1 from pyk595/branch-A-Assertions

Use Assertions

pyk595 committed:
Improve code quality

Individual switch cases in CommandManager class contain repeated
code for saving.

This is a potential source of error in the event of code
deprecation.

Let's refactor code for saving into individual command classes.

Placing save functionality in individual command classes
reduces length of code in the CommandManager::runCommand
method.

pyk595 committed:
Merge pull request #2 from pyk595/branch-A-CodeQuality

Improve code quality

pyk595 committed:
Add Update functionality

Tasks are updated by deleting existing tasks and repopulating
with an updated version of the same task.

This behaviour causes potential loss in information.

Creating an update command allows us to maintain the task objects
in the same list while changing the attributes in one command.

Let's add the update command to allow mutability of task objects.

This behaviour is prefereable because deleting objects and adding
copies of the same object is cumbersome and error prone.

pyk595 committed:
Merge pull request #3 from pyk595/branch-C-Update

Add Update functionality

pyk595 committed:
Set up Continuous Integration (CI)

pyk595 committed:
Fix JavaDoc for NumCommand::Execute

pyk595 committed:
Add User Guide

pyk595 committed:
Set theme jekyll-theme-slate

pyk595 committed:
Update README.md

pyk595 committed:
Update README.md

pyk595 committed:
Update README.md

pyk595 committed:
Update README.md

pyk595 committed:
Set theme jekyll-theme-tactile

pyk595 committed:
Update README.md

pyk595 committed:
Improve the GUI

Currently, the user interface only has text in a vertical box.
To make it more aesthetically pleasing, we need to adjust and
match the asymmetric nature of the conversation.

We add a profile picture and a dialog box flip to emulate
user and virtual assistant interaction.

This helps users differentiate between a command and a response.

pyk595 committed:
add image resources

pyk595 committed:
Merge branch 'branch-A-BetterGui'

Update Gui to have profile pictures

pyk595 committed:
Fix conflict on docs/README.md

pyk595 committed:
Fix formatting for checkstyle

pyk595 committed:
Update README.md

pyk595 committed:
Add Screenshot

Add screenshot for README

pyk595 committed:
Refactor example ui image

pyk595 committed:
Update README.md

pyk595 committed:
Update README.md

pyk595 committed:
Update _config.yml

pyk595 committed:
Update README.md

pyk595 committed:
Rename ui.png

pyk595 committed:
Refactor ui.png

pyk595 committed:
Update README.md

pyk595 committed:
Release Larry Bot v0.2

Create a new release on GitHub

pyk595 committed:
Merge branch 'master' of https://github.com/pyk595/ip
